### PR TITLE
Multiple geth

### PR DESCRIPTION
**Pull request by jeddie-yhc**

### CODE REVIEW

File: cmd/indexer/root.go [hunk 1]
```diff
@@ -204,10 +214,21 @@ func initViper() {
...
 func assignVarFromViper() {
-	// flags for ethereum service
-	ethProtocol = viper.GetString(flags.EthProtocol)
-	ethHost = viper.GetString(flags.EthHost)
-	ethPort = viper.GetInt(flags.EthPort)
+	// flags for eth
+	// eth value has been passed via flag "--eth"
+	ethList := viper.GetStringSlice(flags.Eth)
+	for _, eth := range ethList {
+		u, _ := url.Parse(eth)
```
Comment (yurenju) [inline cmd/indexer/root.go]: should we handle error here? it seems better if it stop executing when url is wrong format

File: configs/config.yaml [hunk 2]
```diff
@@ -1,10 +1,13 @@
 eth:
-  protocol: ws
-  host: geth
-  port: 8546
+  - protocol: ws
```
Comment (markya0616) [inline configs/config.yaml]: How about make flags and configs consistent? Pass array like:
eth:
  - ws://127.0.0.1:8585
  - ws://127.0.0.1:8586

File: cmd/indexer/root.go [hunk 3]
```diff
@@ -204,10 +214,21 @@ func initViper() {
...
+
+		port, _ := strconv.Atoi(u.Port())
+		ethClients = append(ethClients, EthClient{
+			Protocol: u.Scheme,
+			Host:     u.Hostname(),
+			Port:     port,
+		})
+	}
+	// "eth" has been defined in config.yaml
+	viper.UnmarshalKey(flags.Eth, &ethClients)
```
Comment (markya0616) [inline cmd/indexer/root.go]: We should load config first and override by flags

File: docker-compose.yml [hunk 4]
```diff
@@ -1,13 +1,24 @@
...
   # Customized geth image for indexer
-  geth:
+  geth_1:
     image: quay.io/amis/geth:indexer_diff_tx_1.8.10
     ports:
       - "30303:30303"
       - "8546:8546"
     volumes:
-      - ${GETH_DATA_PATH}:/root/.ethereum
+      - /Users/yurenju/go/src/github.com/getamis/geth_1:/root/.ethereum
```
Comment (markya0616) [inline docker-compose.yml]: yurenju?

File: configs/config.yaml [hunk 5]
```diff
@@ -1,10 +1,10 @@
...
+    host: 127.0.0.1
```
Comment (yurenju) [inline configs/config.yaml]: it should be `geth`

File: configs/config.yaml [hunk 6]
```diff
@@ -1,10 +1,10 @@
...
+    port: 8546
 db:
   driver: mysql
-  host: idx-database
+  host: 127.0.0.1
```
Comment (yurenju) [inline configs/config.yaml]: it should be `idx-database`

File: service/indexer/indexer.go [hunk 7]
```diff
@@ -83,40 +91,87 @@ func (idx *indexer) SubscribeErc20Tokens(ctx context.Context, addresses []string
 		}
 	}
 
-	return idx.manager.Init(idx.client)
+	return idx.manager.Init(idx.latestClient)
 }
 
 func (idx *indexer) Listen(ctx context.Context, ch chan *types.Header, fromBlock int64) error {
 	childCtx, cancel := context.WithCancel(ctx)
```
Comment (markya0616) [inline service/indexer/indexer.go]:
```
func (idx *indexer) Listen(ctx context.Context, ch chan *types.Header, fromBlock int64) error {
	// outChannel collects heads from ethClients
	outChannel := make(chan *ethResult)
	idx.wg.Add(len(idx.clients))
	defer idx.wg.Wait()

	// each ethClient will run in different go routines and push messages to channels
	for i, c := range idx.clients {
		go func(index int, client client.EthClient) {
			defer idx.wg.Done()
			childCtx, cancel := context.WithCancel(ctx)
			defer cancel()

			for {
				sub, err := client.SubscribeNewHead(childCtx, ch)
				if err != nil {
					log.Warn("Failed to subscribe event for new header", "client", index, "err", err)
					continue
				}
				// Check whether null exceptions
				defer sub.Unsubscribe()
				for {
					select {
					case head := <-ch:
						result := &ethResult{
							head:        head,
							clientIndex: index,
						}
						outChannel <- result
					case err := <-sub.Err():
						log.Warn("Receive subscribe error", "client", index, "err", err)
						continue
					case <-childCtx.Done():
						return
					}
				}
			}
		}(i, c)
	}
	// single thread to process the messages from channel in sequential
	for {
		select {
		case result := <-outChannel:
			if fromBlock > result.head.Number.Int64() {
				log.Trace("Ignore old header", "number", result.head.Number, "hash", result.head.Hash().Hex())
				continue
			}
			log.Trace("Got new header", "number", result.head.Number, "hash", result.head.Hash().Hex(), "index", result.clientIndex)
			// switch the current ethClient to the source
			idx.latestClient = idx.clients[result.clientIndex]
			err := idx.sync(ctx, fromBlock, result.head.Number.Int64())
			if err != nil {
				log.Error("Failed to sync from ethereum", "number", result.head.Number, "err", err)
				return err
			}
		case <-ctx.Done():
			return childCtx.Err()
		}
	}
}
```

Comment (jeddie-yhc) [reply service/indexer/indexer.go]: 100 restart

Comment (jeddie-yhc) [reply service/indexer/indexer.go]: 1. each ch
2. outChannel add buffer

**Pull request closed without merging**